### DISCUSSION TITLE
modifySelectors in v3

### DISCUSSION THREAD

**Discussion by cwgw** [tailwindlabs/tailwindcss#6691] [Help] [answered]
I noticed that the following example was removed from the "Adding variants" section of the [Plugins documentation](https://tailwindcss.com/docs/plugins#adding-variants) with the upgrade to v3 ([6d6ee63](https://github.com/tailwindlabs/tailwindcss.com/commit/6d6ee63ba619a78955e6e39a46535f80128d839d)). There is no longer any mention of `modifySelectors` in the docs (maybe I missed it?). Is this still a valid/recommended method of adding custom variants?

```javascript
const plugin = require('tailwindcss/plugin')

module.exports = {
  plugins: [
    plugin(function({ addVariant, e }) {
      addVariant('disabled', ({ modifySelectors, separator }) => {
        modifySelectors(({ className }) => {
          return `.${e(`disabled${separator}${className}`)}:disabled`
        })
      })
    })
  ]
}
```

**Comment by alexanderbluhm** [comment, marked as answer]
Yes you can use [the new Variant API](https://tailwindcss.com/docs/plugins#adding-variants):
```js
const plugin = require('tailwindcss/plugin')

module.exports = {
  // ...
  plugins: [
    plugin(function({ addVariant }) {
      addVariant('optional', '&:optional')
      addVariant('hocus', ['&:hover', '&:focus'])
      addVariant('supports-grid', '@supports (display: grid)')
    })
  ]
}
```

[Here](https://github.com/tailwindlabs/tailwindcss/pull/5809) are a few more infos about the new variant API.
However, (fyi @RobinMalfait) it would be nice to have a bit more information directly in the docs or, for the mean time, a link to your explanations on GitHub.

**Reply by RobinMalfait** [reply]
@alexanderbluhm What do you mean with "it would be nice to have a bit more information directly in the docs"?

This should cover everything you need: https://tailwindcss.com/docs/plugins#adding-variants

The PR you linked contains more verbose APIs, but those are not exposed as public API.

**Reply by alexanderbluhm** [reply]
@RobinMalfait Ah okay, I thought that it is exposed as a public API. However, I think `:merge` is usable for the users but not described in the docs, right? In addition, I think that the documentation about adding variants is pretty fine for people like us but a bit too short for Tailwind newcomers.
Nevertheless, we will see how the users deal with it in the discord and in the discussions and then it can be refined it if necessary :)

**Reply by adamwathan** [reply]
Yeah I plan to add more documentation, we were just in a rush to get v3 released on time so some areas had to be more brief to make sure we at least had the basics covered across the whole site.

**Reply by alexanderbluhm** [reply]
No worries I know it was a tight schedule on v3!
